src/libcore/task/wake.rs

```diff
@@ -98,14 +203,32 @@ unsafe impl Sync for Waker {}
 
 impl Waker {
     /// Wake up the task associated with this `Waker`.
-    pub fn wake(&self) {
+    #[inline]
+    pub fn wake(self) {
         // The actual wakeup call is delegated through a virtual function call
         // to the implementation which is defined by the executor.
 
-        // SAFETY: This is safe because `Waker::new_unchecked` is the only way
+        // SAFETY: This is safe because `Waker::from_raw` is the only way
         // to initialize `wake` and `data` requiring the user to acknowledge
         // that the contract of `RawWaker` is upheld.
-        unsafe { (self.waker.vtable.wake)(self.waker.data) }
+        unsafe { (self.waker.vtable.wake)(self.waker.data) };
+
+        // Don't call `drop`-- the waker was already consumed.
+        crate::mem::forget(self);
```

Comment:
This doesn't appear to be panic safe (and even if it's not it will result in worse code). I would expect the fields for the function call to be copied out, then forget called, then the actual call to be done.

Reply:
It's not panic-safe, you're correct. I'm happy to make that change.